### PR TITLE
DS-4412 Create Integration Tests to prove that users in the Administrators group get full Admin Rights

### PR DESCRIPTION
**Pull request by Micheleboychuk**
## References
https://jira.lyrasis.org/browse/DS-4412
## Description
REST API should have basic Integration Tests that prove that a user account added to the Administrator group is given full access rights.
We have tested that the enroll process for a new administrator, i.e. create an user and add him to the administrators group, will result in the Administrator Feature(s) to be available to such user.
## Instructions for Reviewers
We have created a new AuthorizationFeature to verify that an user has access to the administrative feature of the repository or of a specific community and collection.
## Checklist
_This checklist provides a reminder of what we are going to look for when reviewing your PR. You need not complete this checklist prior to creating your PR (draft PRs are always welcome). If you are unsure about an item in the checklist, don't hesitate to ask. We're here to help!_
- [x] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & integration tests). Exceptions may be made if previously agreed upon.
- [x] My PR passes Checkstyle validation based on the [Code Style Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide)
- [x] My PR includes Javadoc for _all new (or modified) public methods and classes_. It also includes Javadoc for large or complex private methods.
- [x] My PR passes all tests and includes new/updated Unit or Integration Tests for any bug fixes, improvements or new features. A few reminders about what constitutes good tests:
* Include tests for different user types, including: (1) Anonymous user, (2) Logged in user (non-admin), and (3) Administrator.
* Include tests for known error scenarios and error codes (e.g. `400 Bad Request`, `401 Unauthorized`, `403 Forbidden`, `404 Not Found`, etc)
* For bug fixes, include a test that reproduces the bug and proves it is fixed. For clarity, it may be useful to provide the test in a separate commit from the bug fix.
- **N/A** If my PR includes new, third-party dependencies (in any `pom.xml`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/master/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- **N/A** If my PR modifies the REST API, I've linked to the REST Contract page (or open PR) related to this change.

### CODE REVIEW

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/authorization/impl/AdministratorFeature.java [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+ * The administrator feature. It can be used for verify that an user has access
+ * to the administrative feature of the repository or of a specific community and collection.
+ *
+ * @author Mykhaylo Boychuk (mykhaylo.boychuk at 4science.it)
+ */
+@Component
+@AuthorizationFeatureDocumentation(name = AdministratorFeature.NAME)
+public class AdministratorFeature implements AuthorizationFeature {
+
+    public static final String NAME = "administratorfeature";
```
Comment (tdonohue) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/authorization/impl/AdministratorFeature.java:36]: We might just want to call this new feature "administratorOf".  It seems to be checking whether you are an Administrator of a particular object (Community, Collection or Site).

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/authorization/EnrollAdministratorIT.java [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+        EPerson eperson1 = EPersonBuilder.createEPerson(context)
+                          .withNameInMetadata("Jhon", "Brown")
+                          .withEmail("johnbrown@example.com")
+                          .withPassword(password)
+                          .build();
+
+        context.restoreAuthSystemState();
+
+        Site site = siteService.findSite(context);
+        SiteRest siteRest = converterService.toRest(site, DefaultProjection.DEFAULT);
```
Comment (tdonohue) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/authorization/EnrollAdministratorIT.java:79]: Just a note. A lot of these Converters get refactored/renamed in the (currently open) PR #2726.  So, depending on which PR is merged first (this one or that one), the other will require refactoring slightly.

**Review by tdonohue — COMMENTED** [review]
@Micheleboychuk : Gave this a quick look today.  This looks good overall, just one minor comment below. 

I also did mention the fact that I see overlap/code conflicts between this PR and #2726.  There's not much that can be done about that *now*, but these PRs may need closer alignment (either that, or once one is merged, the other will need some minor refactoring).

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/authorization/impl/AdministratorOf.java [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+ * The administrator feature. It can be used for verify that an user has access
+ * to the administrative features of the repository or of a specific community and collection.
+ *
+ * @author Mykhaylo Boychuk (mykhaylo.boychuk at 4science.it)
+ */
+@Component
+@AuthorizationFeatureDocumentation(name = AdministratorOf.NAME,
+        description = "It can be used for verify that an user has access "
+                    + "to the administrative features of the repository or of a specific community and collection")
+public class AdministratorOf implements AuthorizationFeature {
```
Comment (tdonohue) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/authorization/impl/AdministratorOf.java:36]: Small clarification to my previous comment.  To align with other existing "Feature" classes, I'd recommend we name the *class* `AdministratorOfFeature`, but keep the `NAME` property below as `administratorOf`.

These two names need not match, and in fact they are named similarly in other Feature classes, e.g. https://github.com/DSpace/DSpace/blob/master/dspace-server-webapp/src/main/java/org/dspace/app/rest/authorization/impl/WithdrawFeature.java or https://github.com/DSpace/DSpace/blob/master/dspace-server-webapp/src/main/java/org/dspace/app/rest/authorization/impl/ReinstateFeature.java

**Review by tdonohue — COMMENTED** [review]
Overall, this PR looks good to me.  However, I needed to clarify a previous request inline below (apologies for the confusion, as it sounds like my prior review wasn't clear)

**Comment by jtimal** [comment]
The tests I did with the different roles were successful, do I have to put the time it takes to do them?

**Comment by tdonohue** [comment]
@jtimal : you do not need to track or report the time it takes for you to do testing or review.  Only *paid* developers (those being paid by LYRASIS to work specifically on DSpace 7) are required to report their time. All other developers/reviewers do not need to report time.

**Review by tdonohue — APPROVED** [review]
👍 This looks good to me now.  Thanks @Micheleboychuk !

**Comment by tdonohue** [comment]
Merging as this is at +2 from Julian and I, and it's primarily ITs.

**Pull request merged**